File: src/CSharp/src/Tools.DataStructures.Hamt/MerkleEncoding.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace Tools.DataStructures.Hamt;

/// <summary>Encodes values into canonical bytes for content-addressed structures.</summary>
/// <typeparam name="T">The encoded value type.</typeparam>
public interface IMerkleCodec<in T>
{
    /// <summary>Gets a stable, versioned identifier for the byte encoding.</summary>
    string EncodingId { get; }

    /// <summary>Encodes one value into a newly owned canonical byte array.</summary>
    /// <param name="value">The value to encode.</param>
    /// <returns>The canonical bytes.</returns>
    byte[] Encode(T value);
}

/// <summary>Provides built-in canonical codecs for common key and value types.</summary>
public static class MerkleCodecs
{
    /// <summary>Gets the signed 32-bit big-endian codec.</summary>
    public static IMerkleCodec<int> Int32 { get; } = new Int32Codec();

    /// <summary>Gets the signed 64-bit big-endian codec.</summary>
    public static IMerkleCodec<long> Int64 { get; } = new Int64Codec();

    /// <summary>Gets the nullable UTF-8 string codec with an explicit null/present tag.</summary>
    public static IMerkleCodec<string?> Utf8String { get; } = new StringCodec();

    /// <summary>Gets the nullable byte-array codec with an explicit null/present tag.</summary>
    public static IMerkleCodec<byte[]?> Bytes { get; } = new BytesCodec();

    /// <summary>Gets the RFC-4122/network-order GUID codec.</summary>
    public static IMerkleCodec<Guid> Guid { get; } = new GuidCodec();

    private sealed class Int32Codec : IMerkleCodec<int>
    {
        public string EncodingId => "i32-be-v1";

        public byte[] Encode(int value)
        {
            var bytes = new byte[sizeof(int)];
            BinaryPrimitives.WriteInt32BigEndian(bytes, value);
            return bytes;
        }
    }

    private sealed class Int64Codec : IMerkleCodec<long>
    {
        public string EncodingId => "i64-be-v1";

        public byte[] Encode(long value)
        {
            var bytes = new byte[sizeof(long)];
            BinaryPrimitives.WriteInt64BigEndian(bytes, value);
            return bytes;
        }
    }

    private sealed class StringCodec : IMerkleCodec<string?>
    {
        private static readonly UTF8Encoding Utf8 = new(false, true);

        public string EncodingId => "nullable-utf8-v1";

        public byte[] Encode(string? value)
        {
            if (value is null)
                return [0];
            var payload = Utf8.GetBytes(value);
            var result = new byte[payload.Length + 1];
            result[0] = 1;
            payload.CopyTo(result, 1);
            return result;
        }
    }

    private sealed class BytesCodec : IMerkleCodec<byte[]?>
    {
        public string EncodingId => "nullable-bytes-v1";

        public byte[] Encode(byte[]? value)
        {
            if (value is null)
                return [0];
            var result = new byte[value.Length + 1];
            result[0] = 1;
            value.CopyTo(result, 1);
            return result;
        }
    }

    private sealed class GuidCodec : IMerkleCodec<Guid>
    {
        public string EncodingId => "guid-rfc4122-v1";

        public byte[] Encode(Guid value)
        {
            Span<byte> little = stackalloc byte[16];
            value.TryWriteBytes(little);
            var result = new byte[16];
            result[0] = little[3];
            result[1] = little[2];
            result[2] = little[1];
            result[3] = little[0];
            result[4] = little[5];
            result[5] = little[4];
            result[6] = little[7];
            result[7] = little[6];
            little[8..].CopyTo(result.AsSpan(8));
            return result;
        }
    }
}

/// <summary>Represents a 256-bit SHA-256 content digest.</summary>
public readonly struct MerkleDigest : IEquatable<MerkleDigest>, IComparable<MerkleDigest>
{
    private readonly ulong _a;
    private readonly ulong _b;
    private readonly ulong _c;
    private readonly ulong _d;

    internal MerkleDigest(ReadOnlySpan<byte> bytes)
    {
        if (bytes.Length != 32)
            throw new ArgumentException("A Merkle digest must contain exactly 32 bytes.", nameof(bytes));
        _a = BinaryPrimitives.ReadUInt64BigEndian(bytes);
        _b = BinaryPrimitives.ReadUInt64BigEndian(bytes[8..]);
        _c = BinaryPrimitives.ReadUInt64BigEndian(bytes[16..]);
        _d = BinaryPrimitives.ReadUInt64BigEndian(bytes[24..]);
    }

    /// <summary>Returns a newly owned 32-byte big-endian digest.</summary>
    /// <returns>The digest bytes.</returns>
    public byte[] ToArray()
    {
        var result = new byte[32];
        WriteTo(result);
        return result;
    }

    /// <summary>Writes the digest to a 32-byte-or-larger destination.</summary>
    /// <param name="destination">The destination span.</param>
    public void WriteTo(Span<byte> destination)
    {
        if (destination.Length < 32)
            throw new ArgumentException("The destination must contain at least 32 bytes.", nameof(destination));
        BinaryPrimitives.WriteUInt64BigEndian(destination, _a);
        BinaryPrimitives.WriteUInt64BigEndian(destination[8..], _b);
        BinaryPrimitives.WriteUInt64BigEndian(destination[16..], _c);
        BinaryPrimitives.WriteUInt64BigEndian(destination[24..], _d);
    }

    /// <inheritdoc/>
    public bool Equals(MerkleDigest other) =>
        _a == other._a && _b == other._b && _c == other._c && _d == other._d;

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is MerkleDigest other && Equals(other);

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(_a, _b, _c, _d);

    /// <inheritdoc/>
    public int CompareTo(MerkleDigest other)
    {
        var comparison = _a.CompareTo(other._a);
        if (comparison != 0) return comparison;
        comparison = _b.CompareTo(other._b);
        if (comparison != 0) return comparison;
        comparison = _c.CompareTo(other._c);
        return comparison != 0 ? comparison : _d.CompareTo(other._d);
    }

    /// <inheritdoc/>
    public override string ToString() => Convert.ToHexString(ToArray()).ToLowerInvariant();

    /// <summary>Tests digest equality.</summary>
    public static bool operator ==(MerkleDigest left, MerkleDigest right) => left.Equals(right);

    /// <summary>Tests digest inequality.</summary>
    public static bool operator !=(MerkleDigest left, MerkleDigest right) => !left.Equals(right);
}

/// <summary>Defines the deterministic comparison, encoding, and hash domain of a Merkle search tree.</summary>
/// <typeparam name="TKey">The key type.</typeparam>
/// <typeparam name="TValue">The value type.</typeparam>
public sealed class MerkleSearchTreePolicy<TKey, TValue>
{
    private static readonly byte[] AlgorithmIdBytes = Encoding.UTF8.GetBytes("mst-sha256-v1");

    private MerkleSearchTreePolicy(
        string policyId,
        IComparer<TKey> comparer,
        IMerkleCodec<TKey> keyCodec,
        IMerkleCodec<TValue> valueCodec)
    {
        PolicyId = policyId;
        Comparer = comparer;
        KeyCodec = keyCodec;
        ValueCodec = valueCodec;
        DomainDigest = HashFramed(
            0x50,
            AlgorithmIdBytes,
            Encoding.UTF8.GetBytes(policyId),
            Encoding.UTF8.GetBytes(keyCodec.EncodingId),
            Encoding.UTF8.GetBytes(valueCodec.EncodingId));
        EmptyDigest = HashFramed(0x00, DomainDigest.ToArray());
    }

    /// <summary>Gets the hash/serialization algorithm identifier.</summary>
    public static string AlgorithmId => "mst-sha256-v1";

    /// <summary>Gets the application-defined semantic policy identifier.</summary>
    public string PolicyId { get; }

    /// <summary>Gets the key comparer.</summary>
    public IComparer<TKey> Comparer { get; }

    /// <summary>Gets the canonical key codec.</summary>
    public IMerkleCodec<TKey> KeyCodec { get; }

    /// <summary>Gets the canonical value codec.</summary>
    public IMerkleCodec<TValue> ValueCodec { get; }

    /// <summary>Gets the digest identifying algorithm, policy, and codec versions.</summary>
    public MerkleDigest DomainDigest { get; }

    internal MerkleDigest EmptyDigest { get; }

    /// <summary>Creates a deterministic Merkle policy.</summary>
    /// <param name="policyId">A stable application semantic/version identifier.</param>
    /// <param name="comparer">The key order/equivalence policy.</param>
    /// <param name="keyCodec">An injective canonical encoding of key equivalence classes.</param>
    /// <param name="valueCodec">A canonical value encoding.</param>
    /// <returns>The policy.</returns>
    public static MerkleSearchTreePolicy<TKey, TValue> Create(
        string policyId,
        IComparer<TKey> comparer,
        IMerkleCodec<TKey> keyCodec,
        IMerkleCodec<TValue> valueCodec)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(policyId);
        ArgumentNullException.ThrowIfNull(comparer);
        ArgumentNullException.ThrowIfNull(keyCodec);
        ArgumentNullException.ThrowIfNull(valueCodec);
        return new(policyId, comparer, keyCodec, valueCodec);
    }

    internal MerkleDigest HashKey(byte[] keyBytes) => HashFramed(0x4b, DomainDigest.ToArray(), keyBytes);

    internal MerkleDigest HashNode(
        byte[] keyBytes,
        byte[] valueBytes,
        MerkleDigest left,
        MerkleDigest right) =>
        HashFramed(0x4e, DomainDigest.ToArray(), keyBytes, valueBytes, left.ToArray(), right.ToArray());

    private static MerkleDigest HashFramed(byte tag, params byte[][] fields)
    {
        using var hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        hash.AppendData([tag]);
        Span<byte> length = stackalloc byte[sizeof(int)];
        foreach (var field in fields)
        {
            BinaryPrimitives.WriteInt32BigEndian(length, field.Length);
            hash.AppendData(length);
            hash.AppendData(field);
        }
        return new MerkleDigest(hash.GetHashAndReset());
    }
}

File: src/Kotlin/Hamt/src/durable7/hamt/MerkleEncoding.kt
package durable7.hamt

import java.nio.ByteBuffer
import java.nio.CharBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.UUID

/** An injective, explicitly versioned canonical value encoding. */
public interface MerkleCodec<T> {
    /** Stable identifier ending in `-v` followed by ASCII decimal digits. */
    public val encodingId: String

    /** Encodes one value into newly owned canonical bytes. */
    public fun encode(value: T): ByteArray

    /** Decodes exactly one complete canonical encoding. */
    public fun decode(encoding: ByteArray): T
}

/** A malformed, noncanonical, or otherwise rejected codec value. */
public class MerkleCodecException(message: String, cause: Throwable? = null) :
    IllegalArgumentException(message, cause)

/** Canonical signed 32-bit big-endian codec (`i32-be-v1`). */
public object Int32MerkleCodec : MerkleCodec<Int> {
    override val encodingId: String = "i32-be-v1"

    override fun encode(value: Int): ByteArray = byteArrayOf(
        (value ushr 24).toByte(),
        (value ushr 16).toByte(),
        (value ushr 8).toByte(),
        value.toByte(),
    )

    override fun decode(encoding: ByteArray): Int {
        if (encoding.size != Int.SIZE_BYTES) {
            throw MerkleCodecException("Invalid $encodingId encoding: expected exactly four bytes.")
        }
        return (encoding[0].toInt() and 0xff shl 24) or
            (encoding[1].toInt() and 0xff shl 16) or
            (encoding[2].toInt() and 0xff shl 8) or
            (encoding[3].toInt() and 0xff)
    }
}

/** Canonical signed 64-bit big-endian codec (`i64-be-v1`). */
public object Int64MerkleCodec : MerkleCodec<Long> {
    override val encodingId: String = "i64-be-v1"

    override fun encode(value: Long): ByteArray = ByteArray(Long.SIZE_BYTES) { index ->
        (value ushr (56 - index * 8)).toByte()
    }

    override fun decode(encoding: ByteArray): Long {
        if (encoding.size != Long.SIZE_BYTES) {
            throw MerkleCodecException("Invalid $encodingId encoding: expected exactly eight bytes.")
        }
        var result = 0L
        for (byte in encoding) result = result shl 8 or (byte.toLong() and 0xff)
        return result
    }
}

/** Canonical nullable strict UTF-8 codec (`nullable-utf8-v1`). */
public object NullableUtf8MerkleCodec : MerkleCodec<String?> {
    override val encodingId: String = "nullable-utf8-v1"

    override fun encode(value: String?): ByteArray {
        if (value == null) return byteArrayOf(0)
        val payload = try {
            val encoder = StandardCharsets.UTF_8.newEncoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)
            val buffer = encoder.encode(CharBuffer.wrap(value))
            ByteArray(buffer.remaining()).also(buffer::get)
        } catch (exception: CharacterCodingException) {
            throw MerkleCodecException("Invalid $encodingId value: input is not valid Unicode.", exception)
        }
        return ByteArray(Math.addExact(payload.size, 1)).also { result ->
            result[0] = 1
            payload.copyInto(result, 1)
        }
    }

    override fun decode(encoding: ByteArray): String? {
        if (encoding.isEmpty()) {
            throw MerkleCodecException("Invalid $encodingId encoding: nullable tag is missing.")
        }
        return when (encoding[0].toInt() and 0xff) {
            0 -> {
                if (encoding.size != 1) {
                    throw MerkleCodecException("Invalid $encodingId encoding: null has trailing bytes.")
                }
                null
            }
            1 -> try {
                val decoder = StandardCharsets.UTF_8.newDecoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT)
                decoder.decode(ByteBuffer.wrap(encoding, 1, encoding.size - 1)).toString()
            } catch (exception: CharacterCodingException) {
                throw MerkleCodecException("Invalid $encodingId encoding: payload is not canonical UTF-8.", exception)
            }
            else -> throw MerkleCodecException(
                "Invalid $encodingId encoding: nullable tag must be zero or one.",
            )
        }
    }
}

/** Canonical nullable byte-array codec (`nullable-bytes-v1`). */
public object NullableBytesMerkleCodec : MerkleCodec<ByteArray?> {
    override val encodingId: String = "nullable-bytes-v1"

    override fun encode(value: ByteArray?): ByteArray {
        if (value == null) return byteArrayOf(0)
        return ByteArray(Math.addExact(value.size, 1)).also { result ->
            result[0] = 1
            value.copyInto(result, 1)
        }
    }

    override fun decode(encoding: ByteArray): ByteArray? {
        if (encoding.isEmpty()) {
            throw MerkleCodecException("Invalid $encodingId encoding: nullable tag is missing.")
        }
        return when (encoding[0].toInt() and 0xff) {
            0 -> {
                if (encoding.size != 1) {
                    throw MerkleCodecException("Invalid $encodingId encoding: null has trailing bytes.")
                }
                null
            }
            1 -> encoding.copyOfRange(1, encoding.size)
            else -> throw MerkleCodecException(
                "Invalid $encodingId encoding: nullable tag must be zero or one.",
            )
        }
    }
}

/** Canonical RFC-4122/network-byte-order UUID codec (`guid-rfc4122-v1`). */
public object Rfc4122UuidMerkleCodec : MerkleCodec<UUID> {
    override val encodingId: String = "guid-rfc4122-v1"

    override fun encode(value: UUID): ByteArray = ByteArray(16).also { result ->
        writeLongBigEndian(result, 0, value.mostSignificantBits)
        writeLongBigEndian(result, 8, value.leastSignificantBits)
    }

    override fun decode(encoding: ByteArray): UUID {
        if (encoding.size != 16) {
            throw MerkleCodecException("Invalid $encodingId encoding: expected exactly sixteen bytes.")
        }
        return UUID(readLongBigEndian(encoding, 0), readLongBigEndian(encoding, 8))
    }
}

/** Immutable 256-bit SHA-256 content address. */
public class MerkleDigest private constructor(private val bytes: ByteArray) : Comparable<MerkleDigest> {
    public companion object {
        public const val BYTE_LENGTH: Int = 32
        public const val HEX_LENGTH: Int = BYTE_LENGTH * 2

        private val sha256: ThreadLocal<MessageDigest> = ThreadLocal.withInitial {
            MessageDigest.getInstance("SHA-256")
        }

        /** Copies exactly 32 digest bytes. */
        public fun fromBytes(bytes: ByteArray): MerkleDigest {
            require(bytes.size == BYTE_LENGTH) {
                "A Merkle digest must contain exactly $BYTE_LENGTH bytes; received ${bytes.size}."
            }
            return MerkleDigest(bytes.copyOf())
        }

        /** Parses exactly 64 hexadecimal characters, accepting either case. */
        public fun fromHex(value: String): MerkleDigest {
            require(value.length == HEX_LENGTH) {
                "A Merkle digest must contain exactly $HEX_LENGTH hexadecimal characters; received ${value.length}."
            }
            val bytes = ByteArray(BYTE_LENGTH)
            for (index in bytes.indices) {
                val high = hexValue(value[index * 2])
                val low = hexValue(value[index * 2 + 1])
                require(high >= 0) { "Invalid hexadecimal character at index ${index * 2}." }
                require(low >= 0) { "Invalid hexadecimal character at index ${index * 2 + 1}." }
                bytes[index] = (high shl 4 or low).toByte()
            }
            return MerkleDigest(bytes)
        }

        /** Computes SHA-256 over exact bytes. */
        public fun hash(bytes: ByteArray): MerkleDigest {
            val digest = sha256.get()
            digest.reset()
            return MerkleDigest(digest.digest(bytes))
        }

        internal fun hashFramed(tag: Int, fields: List<ByteArray>): MerkleDigest {
            val digest = sha256.get()
            digest.reset()
            digest.update(tag.toByte())
            for (field in fields) {
                digest.update(Int32MerkleCodec.encode(field.size))
                digest.update(field)
            }
            return MerkleDigest(digest.digest())
        }

        private fun hexValue(value: Char): Int = when (value) {
            in '0'..'9' -> value.code - '0'.code
            in 'a'..'f' -> value.code - 'a'.code + 10
            in 'A'..'F' -> value.code - 'A'.code + 10
            else -> -1
        }
    }

    /** Copies digest bytes into a newly owned array. */
    public fun toByteArray(): ByteArray = bytes.copyOf()

    /** Writes only when 32 bytes fit at [offset], leaving the destination unchanged on failure. */
    public fun tryWriteBytes(destination: ByteArray, offset: Int = 0): Boolean {
        if (offset < 0 || offset > destination.size - BYTE_LENGTH) return false
        bytes.copyInto(destination, offset)
        return true
    }

    internal fun copyInto(destination: ByteArray, offset: Int): Unit {
        bytes.copyInto(destination, offset)
    }

    internal fun byteAt(index: Int): Int = bytes[index].toInt() and 0xff

    override fun compareTo(other: MerkleDigest): Int {
        for (index in bytes.indices) {
            val comparison = byteAt(index).compareTo(other.byteAt(index))
            if (comparison != 0) return comparison
        }
        return 0
    }

    override fun equals(other: Any?): Boolean = other is MerkleDigest && bytes.contentEquals(other.bytes)

    override fun hashCode(): Int = bytes.contentHashCode()

    override fun toString(): String = buildString(HEX_LENGTH) {
        for (byte in bytes) append("%02x".format(byte.toInt() and 0xff))
    }
}

/** Deterministic comparison, encoding, and hash domain for one Merkle search-tree family. */
public class MerkleSearchTreePolicy<K, V> private constructor(
    public val policyId: String,
    public val comparator: Comparator<K>,
    public val keyCodec: MerkleCodec<K>,
    public val valueCodec: MerkleCodec<V>,
    public val domainDigest: MerkleDigest,
    /** Canonical empty-manifest content address for this policy domain. */
    public val emptyDigest: MerkleDigest,
) {
    public companion object {
        public const val ALGORITHM_ID: String = "mst-sha256-b16-v2"

        /** Creates an explicit deterministic policy. */
        public fun <K, V> create(
            policyId: String,
            comparator: Comparator<K>,
            keyCodec: MerkleCodec<K>,
            valueCodec: MerkleCodec<V>,
        ): MerkleSearchTreePolicy<K, V> {
            require(policyId.isNotBlank()) { "A Merkle policy ID must not be blank." }
            validateEncodingId(keyCodec.encodingId, "key codec")
            validateEncodingId(valueCodec.encodingId, "value codec")
            val domain = MerkleDigest.hashFramed(
                0x50,
                listOf(
                    strictUtf8(ALGORITHM_ID, "algorithm ID"),
                    strictUtf8(policyId, "policy ID"),
                    strictUtf8(keyCodec.encodingId, "key codec ID"),
                    strictUtf8(valueCodec.encodingId, "value codec ID"),
                ),
            )
            val emptyManifest = ByteArray(5 + MerkleDigest.BYTE_LENGTH)
            "MST2".toByteArray(StandardCharsets.US_ASCII).copyInto(emptyManifest)
            emptyManifest[4] = 0
            check(domain.tryWriteBytes(emptyManifest, 5))
            return MerkleSearchTreePolicy(
                policyId,
                comparator,
                keyCodec,
                valueCodec,
                domain,
                MerkleDigest.hash(emptyManifest),
            )
        }

        /** Creates a natural-order deterministic policy. */
        public fun <K : Comparable<K>, V> natural(
            policyId: String,
            keyCodec: MerkleCodec<K>,
            valueCodec: MerkleCodec<V>,
        ): MerkleSearchTreePolicy<K, V> = create(
            policyId,
            Comparator { left, right -> left.compareTo(right) },
            keyCodec,
            valueCodec,
        )

        /** Counts leading zero base-16 digits, yielding a level from zero through 64. */
        public fun level(digest: MerkleDigest): Int {
            var result = 0
            for (index in 0 until MerkleDigest.BYTE_LENGTH) {
                val byte = digest.byteAt(index)
                if (byte and 0xf0 != 0) return result
                result++
                if (byte and 0x0f != 0) return result
                result++
            }
            return result
        }

        private fun validateEncodingId(value: String, field: String): Unit {
            val marker = value.lastIndexOf("-v")
            val valid = value.isNotEmpty() && value.trim() == value && marker > 0 &&
                marker + 2 < value.length && value.substring(marker + 2).all { it in '0'..'9' }
            require(valid) {
                "Invalid $field encoding ID '$value': expected an unpadded ID ending in -v<digits>."
            }
            strictUtf8(value, "$field encoding ID")
        }

        private fun strictUtf8(value: String, field: String): ByteArray = try {
            val encoder = StandardCharsets.UTF_8.newEncoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)
            val buffer = encoder.encode(CharBuffer.wrap(value))
            ByteArray(buffer.remaining()).also(buffer::get)
        } catch (exception: CharacterCodingException) {
            throw IllegalArgumentException("The Merkle $field is not valid Unicode.", exception)
        }
    }

    /** Returns whether algorithm, semantic policy, and both codec domains agree. */
    public fun isCompatibleWith(other: MerkleSearchTreePolicy<K, V>): Boolean =
        domainDigest == other.domainDigest

    /** Computes the policy-bound key digest from canonical encoded bytes. */
    public fun hashKeyBytes(keyBytes: ByteArray): MerkleDigest = MerkleDigest.hashFramed(
        0x4b,
        listOf(domainDigest.toByteArray(), keyBytes),
    )

    /** Canonically encodes and hashes one key. */
    public fun hashKey(key: K): MerkleDigest = hashKeyBytes(keyCodec.encode(key))

    /** Computes SHA-256 over exact bytes. */
    public fun hashBytes(bytes: ByteArray): MerkleDigest = MerkleDigest.hash(bytes)
}

private fun writeLongBigEndian(destination: ByteArray, offset: Int, value: Long): Unit {
    for (index in 0 until Long.SIZE_BYTES) {
        destination[offset + index] = (value ushr (56 - index * 8)).toByte()
    }
}

private fun readLongBigEndian(source: ByteArray, offset: Int): Long {
    var result = 0L
    for (index in 0 until Long.SIZE_BYTES) {
        result = result shl 8 or (source[offset + index].toLong() and 0xff)
    }
    return result
}
